### PR TITLE
Fix DNS resolution memory leak

### LINKED ISSUE #3553: Leak in `DNS::resolve`
Spotted in the SAN build, we appear to have a path where we leak a `uv_getaddrinfo_t`:

https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=41823&view=logs&j=383d248c-4494-5797-d98f-6cef5140601e&t=ae32e15b-0191-5f39-7410-be7e80165871&l=13133

```

47:     #11 0xda4685 in void std::__1::__invoke_void_return_wrapper<void>::__call<asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long)&, unsigned char const*, unsigned long>(asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long)&, unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/__functional_base:348:9
47:     #12 0xda44ae in std::__1::__function::__alloc_func<asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long), std::__1::allocator<asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long)>, void (unsigned char const*, unsigned long)>::operator()(unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/functional:1540:16
47:     #13 0xd9ee70 in std::__1::__function::__func<asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long), std::__1::allocator<asynchost::NodeConnections::register_message_handlers(messaging::Dispatcher<unsigned int>&)::'lambda1'(unsigned char const*, unsigned long)>, void (unsigned char const*, unsigned long)>::operator()(unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/functional:1714:12
47:     #14 0xb75b1b in std::__1::__function::__value_func<void (unsigned char const*, unsigned long)>::operator()(unsigned char const*&&, unsigned long&&) const /usr/lib/llvm-10/bin/../include/c++/v1/functional:1867:16
47:     #15 0xb74ff0 in std::__1::function<void (unsigned char const*, unsigned long)>::operator()(unsigned char const*, unsigned long) const /usr/lib/llvm-10/bin/../include/c++/v1/functional:2473:12
47:     #16 0xb67cfb in messaging::Dispatcher<unsigned int>::dispatch(unsigned int, unsigned char const*, unsigned long) /__w/1/s/build/../src/ds/messaging.h:166:9
47:     #17 0xe8bd6f in messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)::operator()(unsigned int, unsigned char const*, unsigned long) const /__w/1/s/build/../src/ds/messaging.h:252:15
47:     #18 0xe8bc4c in decltype(std::__1::forward<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)&>(fp)(std::__1::forward<unsigned int>(fp0), std::__1::forward<unsigned char const*>(fp0), std::__1::forward<unsigned long>(fp0))) std::__1::__invoke<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)&, unsigned int, unsigned char const*, unsigned long>(messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)&, unsigned int&&, unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/type_traits:3539:1
47:     #19 0xe8b83a in void std::__1::__invoke_void_return_wrapper<void>::__call<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)&, unsigned int, unsigned char const*, unsigned long>(messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)&, unsigned int&&, unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/__functional_base:348:9
47:     #20 0xe8b5d3 in std::__1::__function::__alloc_func<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long), std::__1::allocator<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)>, void (unsigned int, unsigned char const*, unsigned long)>::operator()(unsigned int&&, unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/functional:1540:16
47:     #21 0xe85e45 in std::__1::__function::__func<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long), std::__1::allocator<messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&)::'lambda'(unsigned int, unsigned char const*, unsigned long)>, void (unsigned int, unsigned char const*, unsigned long)>::operator()(unsigned int&&, unsigned char const*&&, unsigned long&&) /usr/lib/llvm-10/bin/../include/c++/v1/functional:1714:12
47:     #22 0xe83e62 in std::__1::__function::__value_func<void (unsigned int, unsigned char const*, unsigned long)>::operator()(unsigned int&&, unsigned char const*&&, unsigned long&&) const /usr/lib/llvm-10/bin/../include/c++/v1/functional:1867:16
47:     #23 0xe83a20 in std::__1::function<void (unsigned int, unsigned char const*, unsigned long)>::operator()(unsigned int, unsigned char const*, unsigned long) const /usr/lib/llvm-10/bin/../include/c++/v1/functional:2473:12
47:     #24 0xe82e50 in ringbuffer::Reader::read(unsigned long, std::__1::function<void (unsigned int, unsigned char const*, unsigned long)>) /__w/1/s/build/../src/ds/ring_buffer.h:148:9
47:     #25 0xe826c4 in messaging::BufferProcessor::read_n(unsigned long, ringbuffer::Reader&) /__w/1/s/build/../src/ds/messaging.h:248:23
47:     #26 0x10eaf25 in asynchost::HandleRingbufferImpl::on_timer() /__w/1/s/build/../src/host/handle_ring_buffer.h:106:10
47:     #27 0x10eadda in asynchost::Timer<asynchost::HandleRingbufferImpl>::on_timer() /__w/1/s/build/../src/host/timer.h:48:17
47:     #28 0x10bc6fa in asynchost::Timer<asynchost::HandleRingbufferImpl>::on_timer(uv_timer_s*) /__w/1/s/build/../src/host/timer.h:43:42
47:     #29 0x7f6224f1d255 in uv__run_timers (/lib/x86_64-linux-gnu/libuv.so.1+0xd255)
47: 
47: Indirect leak of 64 byte(s) in 1 object(s) allocated from:
47:     #0 0x49da5d in malloc (/__w/1/s/build/cchost+0x49da5d)
47:     #1 0x7f622491bd79  (/lib/x86_64-linux-gnu/libc.so.6+0x106d79)
47:     #2 0x7f622491e0d8 in getaddrinfo (/lib/x86_64-linux-gnu/libc.so.6+0x1090d8)
47:     #3 0x44cdaa in getaddrinfo (/__w/1/s/build/cchost+0x44cdaa)
47:     #4 0x7f6224f265fc  (/lib/x86_64-linux-gnu/libuv.so.1+0x165fc)
47: 
47: SUMMARY: AddressSanitizer: 224 byte(s) leaked in 2 allocation(s).
```

This is related to previous libuv issues with `DNS::resolve` (eg #3491), but not identical.

I think the full fix here is likely to be a large rework; promoting `uv_getaddrinfo_t` to a managed type like our `uv_tcp_t` handles, and then following libuv sample flows - you first create a `getaddrinfo`, and _only if that succeeds_ do you go on to create a `tcp` with the resulting info.

### PR DESCRIPTION
**Pull request by wintersteiger**
With hindsight, this was a completely obvious memory leak...

Fixes #3553

### CODE REVIEW

**Review by jumaffre — APPROVED** [review]

**Review by eddyashton — APPROVED** [review]

File: src/host/tcp.h [hunk 1]
```diff
@@ -608,6 +608,12 @@ namespace asynchost
       {
         static_cast<TCPImpl*>(req->data)->on_resolved(req, rc);
       }
+      else
+      {
+        // The TCPImpl that submitted the request has been destroyed, but we
+        // need to clean up the request object.
+        delete req;
```
Comment (achamayou) [inline src/host/tcp.h:615]: @wintersteiger is removing this fixing a double free rather a leak? Am I missing something?

Comment (wintersteiger) [reply src/host/tcp.h:615]: No, it fixes a leak. `req->data` usually holds a pointer to the owning `TCPImpl` and when that gets deleted, the pointers pointing to it a reset to null. The request objects are deleted in the callbacks, but when `req->data` is null we don't call the callbacks anymore, so we need to delete the object here.

Comment (achamayou) [reply src/host/tcp.h:615]: Yep, you're right of course. I'm so tired that I'd interpreted this as code being removed rather than added. Sorry!

Comment (wintersteiger) [reply src/host/tcp.h:615]: No worries, thank for confirming!

**Pull request merged**